x-pack/plugins/observability_solution/entities_data_access/server/services/index_pattern_service.ts

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { SavedObjectsClientContract } from '@kbn/core-saved-objects-api-server';
+import { EntityDefinition } from '@kbn/entities-schema';
+import {
+  ENTITY_BASE_PREFIX,
+  ENTITY_HISTORY,
+  ENTITY_LATEST,
+  SO_ENTITY_DEFINITION_TYPE,
+} from '../../common';
+
+type SchemaVersionString = `v${number}`;
+
+export interface EntityIndexPatternService {
+  indexPattern(options?: Options): string;
+  indexPatternByDefinitionId(definitionId: string, options?: Options): string;
+  indexPatternByType(
+    type: string,
+    options: Options & { soClient: SavedObjectsClientContract }
+  ): Promise<string>;
+}
+
+export interface Options {
+  datasets?: 'all' | typeof ENTITY_LATEST | typeof ENTITY_HISTORY;
+  schemaVersion?: SchemaVersionString | SchemaVersionString[] | 'all';
+}
+
+function translateDatasetsOption(datasets: Options['datasets'], useWildcard = true) {
+  if (!datasets || datasets === 'all') {
+    if (useWildcard) {
+      return ['*'] as const;
+    }
+    return [ENTITY_LATEST, ENTITY_HISTORY] as const;
+  }
+  return [datasets] as const;
+}
+
+function translateSchemaVersionOption(schemaVersion: Options['schemaVersion']) {
+  if (!schemaVersion || schemaVersion === 'all') {
+    return ['*'] as const;
+  }
+
+  return Array.isArray(schemaVersion) ? schemaVersion : [schemaVersion];
+}
+
+async function findAllDefinitionsByType(type: string, soClient: SavedObjectsClientContract) {
+  let total = 0;
+  let page = 1;
+  const definitions = [];
+
+  do {
+    const response = await soClient.find<EntityDefinition>({
+      page,
+      type: SO_ENTITY_DEFINITION_TYPE,
+      filter: `${SO_ENTITY_DEFINITION_TYPE}.attributes.type:(${type})`,
+    });
+
+    definitions.push(...response.saved_objects.map(({ attributes }) => attributes.id));
+    total = response.total;
+    page = page + 1;
+  } while (definitions.length < total);
+
+  return definitions;
+}
+
+export function createIndexPatternService(): EntityIndexPatternService {
+  function indexPattern(options?: Options) {
+    const datasets = translateDatasetsOption(options?.datasets);
+    const schemaVersion = translateSchemaVersionOption(options?.schemaVersion);
+
+    return datasets
+      .map((dataset) =>
+        schemaVersion.map((version) => `.${ENTITY_BASE_PREFIX}.${version}.${dataset}.*`)
```

Comment:
```suggestion
        schemaVersion.map((version) => `.${ENTITY_BASE_PREFIX}.${version}.${dataset}*`)
```

I think you have to remove this dot here, since that dot is only available in the history index names?

Reply:
the pattern would be `.entities.<version>.<dataset>.<definition_id>(.date)?` so while not necessary the dot is harmless in this one

Reply:
No, because it excludes the `latest` indices, which would be `.entities.v1.latest.my-definiftion` (no dot at the end, while the current pattern requires the dot before the wildcard part), no?

Reply:
`.entities.v1.history.my_definition.2024-02-12`
`.entities.v1.latest.my_definition`

I think `.entities.*.*.*` would match both

Reply:
Oh, you're right, sorry 😅